### DISCUSSION TITLE
Classes

### DISCUSSION THREAD

**Discussion by jklwn4** [nidud/asmc#4] [Q&A]
Playing with classes i tried to adapt one of your samples, but it doesn´t run. Strange, it assembles in 32 bit as well, even if there is "RCX" inside. What is wrong with this code?

`; https://www.tutorialspoint.com/cplusplus/cpp_inheritance.htm

include stdio.inc

;.pragma comment(linker,"/defaultlib:msvcrt.lib")

int_t typedef sdword

.class Shape
  width       int_t ?
  height      int_t ?

  setWidth    proc :int_t
  setHeight   proc :int_t
.ends

.class Rectangle : public Shape
  getArea     proc
.ends

.data
vtable RectangleVtbl { { Shape_setWidth, Shape_setHeight }, Rectangle_getArea }
rect   Rectangle { { vtable, 0, 0 } }
Rect   ptr Rectangle rect

;    public Rect

.code

Shape::setWidth proc Width:int_t
  mov [rcx].Shape.width,edx
  ret
Shape::setWidth endp

Shape::setHeight proc Height:int_t
  mov [rcx].Shape.height,edx
  ret
Shape::setHeight endp

Rectangle::getArea proc
  mov eax,[rcx].Shape.width
  mul [rcx].Shape.height
  ret
Rectangle::getArea endp

jk_main proc

;  .new Rect:ptr Rectangle, rect

int 3
  Rect.setWidth(5)
  Rect.setHeight(7)

  mov eax, Rect.getArea()
   
;   printf("Total area: %d\n",  Rect.getArea() )
  .return 0

jk_main endp

end jk_main

`

**Comment by nidud** [comment]
> Playing with classes i tried to adapt one of your samples

I guess this is one of the [mixed samples](https://github.com/nidud/asmc/blob/master/source/test/libc/mixed/mixed.asm) with C++.

> but it doesn´t run. 

Could be the `int 3` line.

> Strange, it assembles in 32 bit as well, even if there is "RCX" inside. 

Most of the code here (LIBC, Asmc, samples and tools) are written for 32/64-bit Windows and Linux. In 32 bit [RCX becomes ecx](https://github.com/nidud/asmc/blob/master/include/libc.inc#L285). However, if the class is not declared as fastcall, or the member functions not defined as fastcall, arguments needs to be loaded into registers:
``` assembly
Rectangle::getArea proc

    ldr rcx,this
    mov eax,[rcx].Shape.width
    mul [rcx].Shape.height
    ret

Rectangle::getArea endp
```

LDR does noting in Win64  (`this` already in RCX), moves `this` from stack in Win32, and moves RDI to RCX in Linux64.

> What is wrong with this code?

The code tag is missing (use the <> option in the menu).

**Comment by jklwn4** [comment]
Thanks - i see, this works. I added "int 3" in order to see in a debugger, which code is generated and how the code flow goes.

Next question: what is the advantage of @ComAlloc (used e.g. in class.asm), why would i need it? And how would i instantiate a COM class, CoCreateInstance or is there a macro or something else?

**Comment by nidud** [comment]
> what is the advantage of @ComAlloc (used e.g. in class.asm), why would i need it?

It simplifies the creation process as oppose to do it manually.

> And how would i instantiate a COM class, CoCreateInstance or is there a macro or something else?

Sometimes you just use [predefined classes](https://github.com/nidud/asmc/blob/master/source/test/Microsoft/msdn/IShellLink/IShellLink.asm#L11). Other times you need to [create your own](https://github.com/nidud/asmc/blob/master/source/test/Microsoft/Samples/DesktopToasts/DesktopToastsSample.asm#L76) to interact with the OS, so it depends on the situation.

**Comment by jklwn4** [comment]
> It simplifies the creation process as oppose to do it manually.
How? I don´t need it in the code of post #1, why do i need it in "class.asm", do i need it at all there? 
The code in post #1 references the class in a global way, so a second reference would be at the same memory location as the first one, is this correct?
Whereas @ComAlloc allocates memory for the class, this way i could have as many independent (not overlapping in memory) instances/references as i like, is this correct?

**Comment by nidud** [comment]
A class is two structures. One for data and one for function pointers. The first element of the class is a pointer to the second structure. When you create a new object you allocate these two, clear the data section and initialize the pointers, so this is what the macro do.


In [class.asm](https://github.com/nidud/asmc/blob/master/source/test/libc/class/class.asm#L41) the class is created by the `.NEW` directive. This calls the default constructor for the class.

In general OOP allocate all necessities on creation and thus try to avoid using any static data. In cases where multiple objects are created only the data structure are allocated as they normally inherit the same functions.

`@ComAlloc(Class[, vtable])
`

**Comment by jklwn4** [comment]
Judging from your examples the default constructor proc must have the same name as the class - is this correct?
Implementing "local" does not call the default constructor, while .new does it - is this correct?
What about the destructor, is this procedure called automatically, when the object goes out of scope, and how would a default destructor be named?

**Comment by nidud** [comment]
> Judging from your examples the default constructor proc must have the same name as the class - is this correct?

Yes, but not part of the class.

> Implementing "local" does not call the default constructor, while .new does it - is this correct?

Yes.

> What about the destructor, is this procedure called automatically, when the object goes out of scope, 

No, you have to [call it](https://github.com/nidud/asmc/blob/master/source/test/Microsoft/Samples/DesktopToasts/DesktopToastsSample.asm#L814C14-L814C14) manually.
 
> and how would a default destructor be named?

It's called Release. Most (if not all) COM objects derives from [IUnknown](https://github.com/nidud/asmc/blob/master/include/unknwn.inc).

``` c
.comdef IUnknown

    QueryInterface	proc :REFIID, :ptr
    AddRef		proc
    Release		proc
   .ends
```

**Comment by jklwn4** [comment]
Some more questions:
what is UNREFERENCED_PARAMETER(xxx)?
i see a declaration of IUNKNOWN, but i must define AddRef, Release and QueryInterface myself - right?
what is the difference beween: lock inc, ...   and InterlockedIncrement ... ? 
what is the difference between: .class and .comdef ?

**Comment by nidud** [comment]
> what is UNREFERENCED_PARAMETER(xxx)?

The debugger display the stack value of parameters and the assembler only add used arguments to the stack. The standard macro (also used in C++) touches these in 64-bit by using `lea rax,args` if \_\_DEBUG\_\_ is defined (set by -Zi).

> i see a declaration of IUNKNOWN, but i must define AddRef, Release and QueryInterface myself - right?

In [some cases](https://github.com/nidud/asmc/blob/master/source/test/Microsoft/Samples/DesktopToasts/DesktopToastsSample.inc#L68) yes, but you may also just use the pointers returned from the OS or other sources.

> what is the difference beween: lock inc, ... and InterlockedIncrement ... ?

The latter are the standard macros defined in [intrin.inc](https://github.com/nidud/asmc/blob/master/include/intrin.inc#L123)

> what is the difference between: .class and .comdef ?

They where different when created but grown more similar over time. The CLASS directive was made to interact with templates and a few other things, but now I think the only difference is use of strict typing of class pointers (`this:ptr[ class]`) and auto alignment.

**Comment by jklwn4** [comment]
Hi nidud.

in the meantime i was able to get to run sample code i had from other languages. So basically it works for me, but there still are questions regarding COM.

- how to retrieve hresult. Reading documentation i learned this is a "hidden" return value telling about error or success, but how to get this value? Some methods in COM interfaces are written to return hresult as return value, but there are e.g. dispatch interfaces returning something else per definition, but these are said to return also hresult - how?
- my next problem comes from a PowerBASIC code sample:

```
' ########################################################################################
' CDO - Sending a text email with an attached file.
' ########################################################################################

#COMPILE EXE
#DIM ALL
#INCLUDE ONCE "CDOSYS.INC"
#INCLUDE ONCE "OLE2UTILS.INC"

' ========================================================================================
' Main
' ========================================================================================
FUNCTION PBMAIN () AS LONG

   LOCAL pMessage AS CDO_IMessage   ' // CDOSYS - IMessage interface
   LOCAL iConf AS CDO_IConfiguration
   LOCAL Flds AS ADOFields

   ' // Create an instance of the CDO Message interface
   pMessage = NEWCOM("CDO.Message")
   IF ISNOTHING(pMessage) THEN EXIT FUNCTION
   iConf = NEWCOM "CDO.Configuration"
   IF ISNOTHING(iConf) THEN EXIT FUNCTION

   TRY
      ' // Configuration
      Flds = iConf.Fields
      LOCAL schema AS STRING
      schema = "http://schemas.microsoft.com/cdo/configuration/"
      Flds.Item(schema & "sendusing").Value = %CdoSendUsingPort
      Flds.Item(schema & "smtpserver").Value = "<your smtp server name>"  ' --> change me
      Flds.Item(schema & "smtpserverport").Value = 25
      Flds.Item(schema & "smtpauthenticate").Value = 1
      Flds.Item(schema & "sendusername").Value = <"your email address">  ' --> change me
      Flds.Item(schema & "sendpassword").Value =  "<your password>"  ' --> change me
      Flds.Item(schema & "smtpusessl").Value = 1
      Flds.Update
      ' // Recipient name --> change as needed
      pMessage.To = "<recipient address>"  ' --> change me
      ' // Sender name --> change as needed
      pMessage.From = "<your email address>"  ' --> change me
      ' // Subject --> change as needed
      pMessage.Subject = "This is a sample subject"
      ' // Text body --> change as needed
      pMessage.TextBody = "This is a sample message text"
      ' // Add the attachment (use absolute paths).
      ' // Note  By repeating the call you can attach more than one file.
      pMessage.AddAttachment "<absolute path of the attachement>"  ' --> change me
      ' // Set the configuration
      pMessage.Configuration = iConf
      ' // Send the message
      pMessage.Send
      ? "Message sent"
   CATCH
      ' // Display the error message
      OleShowErrorInfo OBJRESULT
   FINALLY
      ' // Release the Message object
      pMessage = NOTHING
   END TRY

   #IF %DEF(%PB_CC32)
      WAITKEY$
   #ENDIF

END FUNCTION

```
The interesting parts is this:

> Flds.Item(schema & "smtpusessl").Value = 1  

It "chains" calls, Flds.Item(...)  returns an ADOfield, whose value then is set to 1. How would i do this with asmc. Do i have to split it into two calls or can i do do it in one go (just like in the code sample above)?

Thanks

**Comment by nidud** [comment]
The hresult is just the standard return code for all things COM. There are API's for translating hr to (and from) regular error codes.

Most (if not all) assembler adaption of COM are based on the C implementation which are a bit clumsy compare to the C++ version. Asmc provides the class and template structure for the C++ implementation, so C++ samples are more in sync with Asmc than Basic (and C).

**Comment by jklwn4** [comment]
> The hresult is just the standard return code for all things COM. There are API's for translating hr to (and from) regular error codes.

This is, what i already know, my problem is how to get hresult itself? Sometimes the procedure return (of a method or property) is defined a hresult, then it is clear. But sometimes the returned value is not a hresult, it is some other value or pointer. How (where from) would i get hresult in such a case? 

> ...more in sync with Asmc than Basic (and C).

So there is no asmc equivalent to "Flds.Item(schema & "smtpusessl").Value = 1". I would have to code it in steps, one step for each dot in such syntax constructs, right?

**Comment by nidud** [comment]
Dealing with COM is somewhat repetitive and thus predictable. The chain of events goes something like this:
``` C
    .new foo:ptr IFoo = NULL
    .new type:int_t = 0
    .new hr:HRESULT = ICreate(&foo)
    .if (SUCCEEDED(hr))
        mov hr,foo.GetType(&type)
    .endif
    .if (SUCCEEDED(hr))
        mov hr,...
    .endif
    .if (SUCCEEDED(hr))
        mov hr,...
    .endif
```
In most cases there's a template overlay which include the hr status. This means you interact directly with the pointer throughout the application assuming everything is okay, which imply that there is a built-in exit/exception strategy in the structure.

Example:

``` C
ICreate proto :ptr

.comdef IFoo ; base class

    GetType proc :ptr
   .ends

template macro T
.template T

    nativeType  ptr I&T& ?
    lastResult  HRESULT ?
    scratch     int_t ?

    .static T {
        .if ( !ICreate(addr this.nativeType) )
            ; exit/exception
        .endif
        mov this.lastResult,0
        }

    .static SetStatus watcall :size_t {
        test    eax,eax
        cmovz   eax,this.lastResult ; set if error..
        mov     this.lastResult,eax
        }
    .static GetType {
        mov this.scratch,-1
        this.SetStatus(this.nativeType.GetType(addr this.scratch))
        mov eax,this.scratch
        }
    .static GetStatus {
        mov eax,this.lastResult
        }
    .static GetLastStatus {
        mov eax,this.lastResult
        mov this.lastResult,0
        }
    .ends
    endm

template Foo
``` 
Now hr is removed from the call sequence and embedded in the template.
``` 
    .new p:Foo()

    .if ( p.GetType() == 3 )

        .return( p.GetStatus() )
    .endif
``` 
So most likely there's a member function for this. A good method to find out is to use the VS debugger which will trace these call from the source.

> So there is no asmc equivalent to "Flds.Item(schema & "smtpusessl").Value = 1". I would have to code it in steps, one step for each dot in such syntax constructs, right?

The Asmc syntax will be something like this:
    
    mov [Flds.Item(&schema, "smtpusessl")].Value,1

**Comment by jklwn4** [comment]
> The Asmc syntax will be something like this:
> 
> mov [Flds.Item(&schema, "smtpusessl")].Value,1
> 

So i can indeed "chain" such calls by enclosing each "call" into square brackets? How many nesting levels can i have? 

Would this be possible (pseudo code) too?

`mov [[object1.item(2)].value("asdf")].size, 2 `

**Comment by nidud** [comment]
> So i can indeed "chain" such calls by enclosing each "call" into square brackets?

It's technically not "chained" in the first sample as it is just a function call, but it is implied from the extension that the call returns an address to a struct with a "Value" member.

> How many nesting levels can i have?

There's no nesting limit but "nesting" implies function calls as argument to a function.

    rbx( "first", rdi( "second", rax( "third" ) ) )

    *    lea rcx,DS0000
    *    call rax
    *    lea rcx,DS0001
    *    mov rdx,rax
    *    call rdi
    *    lea rcx,DS0002
    *    mov rdx,rax
    *    call rbx

> Would this be possible (pseudo code) too?
    mov [[object1.item(2)].value("asdf")].size, 2 

Well, there are no nesting here but two function calls in a row, so lets see what happens. I assume something like the following:

``` C++
.class I

    size int_t ?
    name string_t ?

    value proc :string_t
   .ends

.class O

    obj I 20 dup(<>)

    item proc :int_t
   .ends
```

The return from the first call is RAX which is also used to render the call, so ASSUME may not be used here.

    .new o:ptr O

     mov [[o.item(2)].I.value("asdf")].I.size, 2

    *    mov edx, 2
    *    mov rcx, o
    *    mov rax, [rcx]
    *    call [rax].OVtbl.item

Now we have this situation:

    mov [[rax].I.value("asdf")].I.size, 2

    *    lea rdx, DS0000
    *    mov rcx, rax
    *    mov rax, [rcx]
    *    call [rax].IVtbl.value
    *    mov [rax].I.size, 2

So yes, that seems to work.

**Reply by jklwn4** [reply]
Thanks so far for your support, i managed to solve my problems with objects.
Next is events, you can "subscribe" to COM events (which in effect works similar to a callback), but i cannot find ASMC sample code. Did i miss something, or didn´t you do it yourself?

**Comment by nidud** [comment]
> what is the difference between: .class and .comdef ?

Turns out the main difference is that .COMDEF defaults to STDCALL in 32-bit.

> Next is events, you can "subscribe" to COM events (which in effect works similar to a callback), but i cannot find ASMC sample code. Did i miss something, or didn´t you do it yourself?

There are a few of them.

This [CApplication](https://github.com/nidud/asmc/tree/master/source/test/graph/CApplication/wrl) sample shows how to overlay a COM Event class.
```
.class CApplication : public IAppVisibilityEvents

    m_hInstance     HINSTANCE ?
    m_hwnd          HWND ?
    m_width         int_t ?
    m_height        int_t ?
    m_refcnt        int_t ?
    m_launchcnt     int_t ?
    m_cookie        uint_t ?
    m_AppVisibility ptr IAppVisibility ?
    m_yoffs         uint_t ?

    CApplication    proc :HINSTANCE
    Run             proc
    OnKeyDown       proc :WPARAM
    OnClose         proc
    OnDestroy       proc
    OnPaint         proc
    OnSize          proc :LPARAM
    Update          proc :BOOL

    BeforeEnteringMessageLoop   proc
    EnterMessageLoop            proc
    AfterLeavingMessageLoop     proc

    CreateApplicationWindow     proc
    ShowApplicationWindow       proc
    DestroyApplicationWindow    proc
   .ends
```
This receives a notice when the Start button is pressed and auto exits after 5 times. For this to work in 32-bit the calling convention should be set there:
```
.class CApplication WINAPI : public IAppVisibilityEvents
```
It is however exclusively written for 64-bit so that wont help. A more complicated one is the [DesktopDuplication](https://github.com/nidud/asmc/tree/master/source/test/Microsoft/msdn/DesktopDuplication)  sample which now also work in 32-bit (where I found out the difference between .COMDEF and .CLASS).

**Reply by jklwn4** [reply]
Thanks for the info, but my questions wasn´t about Windows events. It was about COM event handling implementing IConnectionPoint et. al. so that a COM object could send "events" to a client. That is, the client exposes methods a COM object can call in oder to inform the client about e.g a mouse click, progress or completion of a task the object is about to process.

**Comment by nidud** [comment]
Sending events is a bit more complicated as you need to install the app by adding a "App.lnk" and a CLSID to the registry.

The sample [DesktopToasts](https://github.com/nidud/asmc/tree/master/source/test/Microsoft/Samples/DesktopToasts) do that (a [Windows-classic-sample](https://github.com/microsoft/Windows-classic-samples/tree/main/Samples/DesktopToasts)).

**Comment by jklwn4** [comment]
In the meantime i managed to bind COM event interfaces via IConnectionPoint. When the COM server "throws" an event, i end up in IDispatch::Invoke. This requires specific code for this procdure in my COM client code.

Given a procedure definition like so

>IDispatch::Invoke proc dispid:sdword, riid:ptr, lcid:dword, wFlags:WORD, params:ptr, result:ptr, pexcepinfo:ptr, puArgErr:ptr

i normally would retrive a ITypeInfo from the corresponding typelib (can be retrieved from riid) and then let "DispInvoke" (a specialized Windows API for this, or "Invoke", a method of ITypeInfo) do all the work for me. Unfortunately this works only for "dual" interfaces (the typelib must have a "dual" attribute). In cases where there is no dual attribute, this method fails and i must resort to do it all by myself. E.g "DWebBrowserEvents2" is such an interface.

I found code in Bitriders ObjAsm i could port:

> ;Push pointer to the ret value (only if DISPATCH_METHOD or DISPATCH_PROPERTYGET are specified)
  mov esi, this
  mov ax, wFlags                      ;.ifBitSet wFlags, (%DISPATCH_PROPERTYGET or %DISPATCH_METHOD)    2 or 1
  and ax, 3
  test ax, ax
  je @F
    mov eax, result
    test eax, eax
    je @F                             ;.if eax != NULL  ;Check if we have a valid pointer" + $crlf + _ 
      mov [eax].VARIANT.w1, 3         ;  VT_I4
      lea ecx, [eax].VARIANT.d2
      push ecx                        ;push return on the stack
@@:
;Extract the arguments from the VARIANTS and push them onto the stack
  mov eax, params
  mov ecx, [eax].DISPPARAMS.cArgs
  mov edx, [eax].DISPPARAMS.rgvarg
  .while ecx != 0
    mov ax, [edx].VARIANT.w1          ;vt
    .if (ax & 00004000h)              ;VT_BYREF
      push [edx].VARIANT.d2           ;push pointer (byref value)
    .elseif eax == 12                 ;VT_VARIANT
      push [edx + 12]
      push [edx + 8]
      push [edx + 4]
      push [edx]                      ;push variant itself
    .elseif (ax == 5 || ax == 7 || ax == 6 || ax == 64)     ;(ax == VT_R8 || ax == VT_DATE || ax == VT_CY || ax == VT_FILETIME)
      push [edx + 12]
      push [edx + 8]                  ;push a 64 bit value
    .else
      push [edx].VARIANT.d2           ;push a 32 bit value
    .endif
    add edx, sizeof VARIANT
    dec ecx
  .endw
  push esi                            ;push "this"
mov eax, _pcall                       ;procedure pointer
call eax                              ;do the call

this builds the stack for a 32 bit stdcall from the variants received ans does the call to the correct procedure, as first tests show. I´m not quite sure ATM, if byte and word arguments are properly handled, but everything not working as expected could be fixed. Basically the skeleton is working and i understand the concept. 

...so, how to do this in 64 bit?

Building the stack for a 32 bit stdcall by hand isn´t that hard, but building the stack for a 64 bit fastcall by hand is much more difficult - and Bitrider didn´t do it too. Of course i could write a separate call for each event procedure involved, but i would prefer to have a more generic approach, like the one above for 32 bit using a procedure pointer.

Any ideas ? 

Getting the correct pointer from the dispid is not a problem, i know, if there is a return value and i know the number of parameters and their types (all types a variant supports). I don´t need strict type checking of parameters, because all of this is declared elsewhere in detail and i can rely on the server to pass the correct types, i think.

**Reply by nidud** [reply]
> Any ideas ?

Maybe just return E_NOTIMPL ?

**Comment by nidud** [comment]
When implemented you call members based on `dispIdMember`
```
.if ( pDispParams )
    .switch dispIdMember
    .case DISPID_1
        DispId1(pDispParams.rgvarg[0].lVal, ...
```

**Reply by jklwn4** [reply]
Yes, i didn´t paste the full code, that´s basically what i do. Calling the correct procedure is not the problem, it´s about the parameters.
 
After a little break and thinking about it, i´m going to try building a stack for 64 bit along the lines of my working 32 bit version.

**Reply by nidud** [reply]
> Calling the correct procedure is not the problem, it´s about the parameters.

Well, then you know the arguments.

> i´m going to try building a stack for 64 bit along the lines of my working 32 bit version.

Please don't. Just used the arguments supplied in the structure and make the call.
```
    mov rax,pDispParams
    mov rbx,[rax].DISPPARAMS.rgvarg

    DispCmd( [rbx+VARIANT*0].VARIANT.lVal, [rbx+VARIANT*1].VARIANT.pcVal )
```

**Reply by jklwn4** [reply]
You cannot just pass the values in those variants, you must interpret the content, before you know, what to pass (e.g. it could be "byref" or it could be "byval"). And it´s not only one procedure, it´s a whole lot of event procedures to call. Of course i´m going to build the stack according to 64 bit calling convention rules.

In the meantime, i have some code for testing, will report as soon as i will find time for testing, maybe tomorrow ...

**Reply by nidud** [reply]
> you must interpret the content, before you know, what to pass (e.g. it could be "byref" or it could be "byval").

There's no need to dig into that as you already have all the information you need.

> > Calling the correct procedure is not the problem, it´s about the parameters.
>
> Well, then you know the arguments.
```
DispCmd proto :LONG, :LPSTR 
```
Use the [VARIANT](../blob/master/include/oaidl.inc#L142) struct to select a fitting size:
```
DispCmd( [rbx+VARIANT*0].VARIANT.lVal, [rbx+VARIANT*1].VARIANT.pcVal )
```

**Reply by jklwn4** [reply]
No time for nothing this week ...

this is my (working) 64 bit code:

>   xor xcx, xcx
  xor r11, r11
  mov ax, wFlags             ;wFlags, (%DISPATCH_PROPERTYGET or %DISPATCH_METHOD)    2 or 1
  and ax, 3
  test ax, ax
  je @F
    mov xax, result
    test xax, xax
    je @F                                           ;.if eax != NULL                                   ;Check if we have a valid pointer
      mov xcx, 1
      mov r11, 1
@@:
;get # of arguments to set/push
  mov xax, params
  add ecx, [xax].DISPPARAMS.cArgs
  mov xbx, 020h
  mov r10, xbx
  cmp xcx, 3
  jbe do_reg
    mov xbx, xcx
    mov r10, xcx
    jp @F
      add xbx, 1
      sub rsp, 8
@@:
    shl xbx, 3
    shl r10, 3
;*** actually set and push ***
  test r11, r11      ;return expected ?
  je @F
    mov xax, result
    mov [xax].VARIANT.w1, 20                       ;  VT_I8
    lea xax, qword ptr [xax].VARIANT.d1
    push xax
    sub xcx, 1
    xor r11, r11
@@:
cmp xcx, 3
jbe do_reg
  mov xax, params
  mov xdx, [xax].DISPPARAMS.rgvarg
  .while xcx > 3
    mov ax, [xdx].VARIANT.w1                               ;vt
    .if (ax & 00004000h)                                   ;VT_BYREF
      push qword ptr [xdx].VARIANT.d1                      ; byref
    .elseif ax == 12                                       ;VT_VARIANT
      lea xax, [xdx].VARIANT
      push xax
    .elseif (ax == 5 || ax == 7 || ax == 6 || ax == 64)     ;(ax == VT_R8 || ax == VT_DATE || ax == VT_CY || ax == VT_FILETIME)
      push qword ptr [xdx + 8]
    .else
      push qword ptr [xdx].VARIANT.d1                                 ;lVal
    .endif
    add xdx, sizeof VARIANT
    dec xcx
  .endw
;**********************************************************************
do_reg:
sub xsp, r10
mov xax, params
mov xdi, [xax].DISPPARAMS.rgvarg
.while xcx !=0
  test r11, r11
  je @F
    mov xax, result
      mov [xax].VARIANT.w1, 20                       ;  VT_I8
    .switch xcx
      .case 3: lea r9, qword ptr [xax].VARIANT.d1
      .case 2: lea r8, qword ptr [xax].VARIANT.d1
      .case 1: lea xdx, qword ptr [xax].VARIANT.d1
    .endsw
    xor r11, r11
    sub xcx, 1
    jmp do_next
@@:
    mov ax, [xdi].VARIANT.w1                               ;vt
    .if (ax & 00004000h)                                   ;VT_BYREF
      .switch xcx
        .case 3: lea r9, qword ptr [xdi].VARIANT.d1
        .case 2: lea r8, qword ptr [xdi].VARIANT.d1
        .case 1: lea xdx, qword ptr [xdi].VARIANT.d1
      .endsw
    .elseif ax == 12                                       ;VT_VARIANT
;       lea xax, [xdi].VARIANT
      .switch xcx
        .case 3: lea r9, [xdi].VARIANT
        .case 2: lea r8, [xdi].VARIANT
        .case 1: lea xdx, [xdi].VARIANT
      .endsw
    .elseif (ax == 7 || ax == 6 || ax == 64)     ;(ax == VT_R8 || ax == VT_DATE || ax == VT_CY || ax == VT_FILETIME)
      .switch xcx
        .case 3: mov r9, qword ptr [xdi].VARIANT.d1
        .case 2: mov r8, qword ptr [xdi].VARIANT.d1
        .case 1: mov xdx, qword ptr [xdi].VARIANT.d1
      .endsw
    .elseif (ax == 5)   ;(ax == VT_R8)
      .switch xcx
        .case 3: movq xmm3, qword ptr [xdi].VARIANT.d1
        .case 2: movq xmm2, qword ptr [xdi].VARIANT.d1
        .case 1: movq xmm1, qword ptr [xdi].VARIANT.d1
      .endsw
    .elseif (ax == 4)     ;(ax == VT_R4)
      .switch xcx
        .case 3: movd xmm3, dword ptr [xdi].VARIANT.d2
        .case 2: movd xmm2, dword ptr [xdi].VARIANT.d2
        .case 1: movd xmm1, dword ptr [xdi].VARIANT.d2
      .endsw
    .else
      .switch xcx
        .case 3: mov r9, qword ptr [xdi].VARIANT.d1
        .case 2: mov r8, qword ptr [xdi].VARIANT.d1
        .case 1: mov xdx, qword ptr [xdi].VARIANT.d1
      .endsw
    .endif
    add xdi, sizeof VARIANT
    dec xcx
do_next:
  .endw
;int 3
  mov xcx, this
  mov xax, _pcall
  call xax
  add xsp, xbx
  ret

i use these options:

>   OPTION STACKBASE : RBP
  option win64:5
  option cstack:on            ;save nonvolatile regs before locals
  OPTION FRAME:AUTO           ;SEH

this allows for a stackframe like in 32 bit, the only thing i must keep an eye on is having RSP aligned to 16 bytes. I know this is not the recommended way for 64 bit, but for serveral reasons i want it this way and (most important) it actually works this way. 

My VARIANT struct looks different too (i don´t want to explain here why, please just take it as it is):

> VARIANT struc
  w1 word ?                                           ;.vt
  w2 word ?                                           ;.reserved1
  w3 word ?                                           ;.reserved2
  w4 word ?                                           ;.reserved3
  d1 dword ?                                          ;data union
  d2 dword ?
VARIANT ends

**Reply by nidud** [reply]
> My VARIANT struct looks different

The size should at least be corrected for the API functions that use it.

> i use these options:

These options do not have any meaning without the AUTO (bit 1) option. The STACK calculation is done by the PROC directive based on the INVOKE directives used between PROC and ENDP.

> mov xax, _pcall

You should invoke this directly instead of doing all of this.
```
this._pcall(...)
```
